ktor-server/ktor-server-core/jvm/src/io/ktor/application/newapi/Interceptions.kt

```diff
@@ -0,0 +1,165 @@
+/*
+ * Copyright 2014-2021 JetBrains s.r.o and contributors. Use of this source code is governed by the Apache 2.0 license.
+ */
+
+package io.ktor.application.newapi
+
+import io.ktor.application.*
+import io.ktor.request.*
+import io.ktor.response.*
+import io.ktor.util.*
+import io.ktor.util.pipeline.*
+import kotlin.random.*
+
+/**
+ * Compatibility class. It describes how (with what new functionality) some particular phase should be intercepted.
+ * It is a wrapper over pipeline.intercept(phase) { ... } and is needed to hide old Plugins API functionality
+ * */
+public class Interception<T : Any>(
+    public val phase: PipelinePhase,
+    public val action: (Pipeline<T, ApplicationCall>) -> Unit
+)
+
+/**
+ * Compatibility class. Interception class for Call phase
+ * */
+public typealias CallInterception = Interception<Unit>
+
+/**
+ * Compatibility class. Interception class for Receive phase
+ * */
+
+public typealias ReceiveInterception = Interception<ApplicationReceiveRequest>
+
+/**
+ * Compatibility class. Interception class for Send phase
+ * */
+public typealias ResponseInterception = Interception<Any>
+
+/**
+ * Compatibility class. Every plugin that needs to be compatible with `beforePlugin(...)` and `afterPlugin(...)`
+ * needs to implement this class. It defines a list of interceptions (see [Interception]) for phases in different pipelines
+ * that are being intercepted by the current plugin.
+ *
+ * It is needed in order to get first/last phase in each pipeline for the current plugin and create a new phase
+ * that is strictly earlier/later in this pipeline than any interception of the current plugin.
+ *
+ * Note: any [KtorPlugin] instance automatically fills these fields and there is no need to implement them by hand.
+ * But if you want to use old plugin API based on pipelines and phases and to make it available for `beforePlugin`/
+ * `afterPlugin` methods of the new plugins API, please consider filling the phases your plugin defines by implementing
+ * this interface [InterceptionsHolder].
+ *
+ * Please, use [defineInterceptions] method inside the first init block to simply define all the interceptions for your
+ * feature. Also it is useful to always delegate to [DefaultInterceptionsHolder] for simplicity.
+ * */
+public interface InterceptionsHolder {
+    /**
+     * A name for the current plugin.
+     * */
+    public val name: String get() = this.javaClass.simpleName
+
+    public val key: AttributeKey<out InterceptionsHolder>
+
+    @InternalAPI
```

Comment:
@InternalAPI will be dropped in 2.0.0

Reply:
@e5l why?

Reply:
What is the alternative?

Reply:
there is no alternative, all API either stable or private